### PR TITLE
Fixes #1207: Disabling WPML via WPML Network Setup causes PHP Fatal Error on Multisite

### PR DESCRIPTION
**Pull request by Screenfeed**
Also improved lots of things.

### CODE REVIEW

File: inc/common/admin-bar.php [hunk 1]
```diff
@@ -12,299 +12,300 @@
...
 					'title'  => __( 'Clear cache', 'rocket' ),
 					'href'   => '#',
-				)
+				]
 			);
 
-			$langlinks = get_rocket_wpml_langs_for_admin_bar();
-			if ( $langlinks ) {
+			// Add submenu for each active langs.
+			if ( 'wpml' === $i18n_plugin ) {
```
Comment (Tabrisrp) [inline inc/common/admin-bar.php]: Maybe a switch statement instead of the multiple elseif

File: inc/common/admin-bar.php [hunk 2]
```diff
@@ -12,299 +12,300 @@
...
-			foreach ( $langlinks as $lang ) {
-				$wp_admin_bar->add_menu(
-					array(
-						'parent' => 'purge-all',
-						'id'     => 'purge-all-' . $lang['code'],
-						'title'  => $lang['flag'] . '&nbsp;' . $lang['anchor'],
-						'href'   => wp_nonce_url( admin_url( 'admin-post.php?action=' . $action . '&type=all&lang=' . $lang['code'] . $referer ), $action . '_all' ),
-					)
-				);
+				if ( 'wpml' !== $i18n_plugin ) {
```
Comment (Tabrisrp) [inline inc/common/admin-bar.php]: this should be before the previous conditional, to have the "All" link first in the list

File: inc/common/admin-bar.php [hunk 3]
```diff
@@ -12,299 +12,300 @@
...
-				foreach ( $langlinks as $lang ) {
-					$wp_admin_bar->add_menu(
-						array(
-							'parent' => 'preload-cache',
-							'id'     => 'preload-cache-' . $lang['code'],
-							'title'  => $lang['flag'] . '&nbsp;' . $lang['anchor'],
-							'href'   => wp_nonce_url( admin_url( 'admin-post.php?action=' . $action . '&lang=' . $lang['code'] . $referer ), $action ),
-						)
-					);
+					if ( 'wpml' !== $i18n_plugin ) {
```
Comment (Tabrisrp) [inline inc/common/admin-bar.php]: Same as above

File: inc/functions/i18n.php [hunk 4]
```diff
@@ -305,113 +329,138 @@ function get_rocket_i18n_to_preserve( $current_lang ) {
...
-		return false;
+	$i18n_plugin = rocket_has_i18n();
+
+	if ( ! $i18n_plugin ) {
+		return [];
 	}
 
-	$urls = array();
-	if ( rocket_is_plugin_active( 'sitepress-multilingual-cms/sitepress.php' ) ) {
+	if ( 'wpml' === $i18n_plugin ) {
```
Comment (Tabrisrp) [inline inc/functions/i18n.php]: a switch here too?

File: inc/functions/i18n.php [hunk 5]
```diff
@@ -305,113 +329,138 @@ function get_rocket_i18n_to_preserve( $current_lang ) {
...
+ * @param  int    $post_id   Post ID.
+ * @param  string $post_type Post Type.
+ * @param  string $regex     Regex to include at the end.
+ * @return array
  */
 function get_rocket_i18n_translated_post_urls( $post_id, $post_type = 'page', $regex = null ) {
-	$urls  = array();
-	$path  = rocket_extract_url_component( get_permalink( $post_id ), PHP_URL_PATH );
-	$langs = get_rocket_i18n_code();
+	$path = rocket_extract_url_component( get_permalink( $post_id ), PHP_URL_PATH );
```
Comment (Tabrisrp) [inline inc/functions/i18n.php]: You can use `wp_parse_url()` directly now

File: inc/functions/i18n.php [hunk 6]
```diff
@@ -305,113 +329,138 @@ function get_rocket_i18n_to_preserve( $current_lang ) {
...
 	// WPML.
-	if ( rocket_is_plugin_active( 'sitepress-multilingual-cms/sitepress.php' ) ) {
-		foreach ( $langs as $lang ) {
-			$urls[] = rocket_extract_url_component( get_permalink( icl_object_id( $post_id, $post_type, true, $lang ) ), PHP_URL_PATH ) . $regex;
+	if ( 'wpml' === $i18n_plugin && $langs ) {
+		$langs = get_rocket_i18n_code();
+
+		if ( $langs ) {
+			foreach ( $langs as $lang ) {
+				$urls[] = rocket_extract_url_component( get_permalink( icl_object_id( $post_id, $post_type, true, $lang ) ), PHP_URL_PATH ) . $regex;
```
Comment (Tabrisrp) [inline inc/functions/i18n.php]: `wp_parse_url()`here too

File: inc/functions/i18n.php [hunk 7]
```diff
@@ -305,113 +329,138 @@ function get_rocket_i18n_to_preserve( $current_lang ) {
...
 		}
 	}
 
 	// qTranslate & qTranslate-x.
-	if ( rocket_is_plugin_active( 'qtranslate/qtranslate.php' ) || rocket_is_plugin_active( 'qtranslate-x/qtranslate.php' ) ) {
+	if ( 'qtranslate' === $i18n_plugin || 'qtranslate-x' === $i18n_plugin ) {
 		$langs  = $GLOBALS['q_config']['enabled_languages'];
 		$langs  = array_diff( $langs, array( $GLOBALS['q_config']['default_language'] ) );
-		$url    = get_permalink( $post_id );
 		$urls[] = rocket_extract_url_component( get_permalink( $post_id ), PHP_URL_PATH ) . $regex;
```
Comment (Tabrisrp) [inline inc/functions/i18n.php]: and here

File: inc/functions/i18n.php [hunk 8]
```diff
@@ -305,113 +329,138 @@ function get_rocket_i18n_to_preserve( $current_lang ) {
...
-			if ( rocket_is_plugin_active( 'qtranslate/qtranslate.php' ) ) {
-				$urls[] = rocket_extract_url_component( qtrans_convertURL( $url, $lang, true ), PHP_URL_PATH ) . $regex;
-			} elseif ( rocket_is_plugin_active( 'qtranslate-x/qtranslate.php' ) ) {
-				$urls[] = rocket_extract_url_component( qtranxf_convertURL( $url, $lang, true ), PHP_URL_PATH ) . $regex;
+		if ( $langs ) {
+			$url = get_permalink( $post_id );
+
+			foreach ( $langs as $lang ) {
+				if ( 'qtranslate' === $i18n_plugin ) {
+					$urls[] = rocket_extract_url_component( qtrans_convertURL( $url, $lang, true ), PHP_URL_PATH ) . $regex;
```
Comment (Tabrisrp) [inline inc/functions/i18n.php]: well everywhere in fact

File: inc/functions/i18n.php [hunk 9]
```diff
@@ -305,113 +329,138 @@ function get_rocket_i18n_to_preserve( $current_lang ) {
...
+				} elseif ( 'qtranslate-x' === $i18n_plugin ) {
+					$urls[] = rocket_extract_url_component( qtranxf_convertURL( $url, $lang, true ), PHP_URL_PATH ) . $regex;
```
Comment (Tabrisrp) [inline inc/functions/i18n.php]: last one

File: inc/common/admin-bar.php (hunk 2 above)
Comment (Screenfeed) [reply inc/common/admin-bar.php]: It was at the end of list until now, do you want me to change it?

**Pull request merged**